### ISSUE TITLE
error connecting a publisher to the connected factory

### ISSUE THREAD

**Issue by eoursel** [Azure/azure-iot-connected-factory#32]
Hi, 

I have a publisher running on a docker container (image from docker hub) on a Centos virtual machine running on my laptop. The publisher is working and connected to an OPC UA server. With iothub-explorer i see that the notifications are correctly sent to IOTHUB. With an OPC UA client, i can connect the Publisher.  
  
Nevertheless when i try to add the publisher to the connected factory dashboard, it says it can't connect  to the OPC UA server. I activated the logs in the web application on azure and there is this exception.  

The connected factory has been deployed automatically by using azureiotsuite.com

Did i miss something?  

Regards

> 2018-06-14T17:00:54  PID[4264] Error       Could not fetch endpoints from url: opc.tcp://publisher:62222/
> 
> 2018-06-14T17:00:54  PID[4264] Error       Reason = Error establishing a connection: BadNotConnected
> 
> 2018-06-14T17:00:54  PID[4264] Error       Exception while trying to connect to the OPC UA server.
> 
> Exceptionmessage:
> 
> Error establishing a connection: BadNotConnected
> 
> Inner Message:
> 
> --
> 
> Stack trace:
> 
>    at Opc.Ua.Bindings.ChannelAsyncOperation`1.End(Int32 timeout, Boolean throwOnError)
> 
>    at Opc.Ua.Bindings.UaSCUaBinaryClientChannel.EndSendRequest(IAsyncResult result)
> 
>    at Opc.Ua.Bindings.UaSCUaBinaryTransportChannel.EndSendRequest(IAsyncResult result)
> 
>    at Opc.Ua.Bindings.UaSCUaBinaryTransportChannel.SendRequest(IServiceRequest request)
> 
>    at Opc.Ua.DiscoveryClient.GetEndpoints(RequestHeader requestHeader, String endpointUrl, StringCollection localeIds, StringCollection profileUris, EndpointDescriptionCollection& endpoints)
> 
>    at Opc.Ua.DiscoveryClient.GetEndpoints(StringCollection profileUris)
> 
>    at Microsoft.Azure.IoTSuite.Connectedfactory.WebApp.Helpers.OpcSessionHelper.DiscoverEndpoints(ApplicationConfiguration config, Uri discoveryUrl, Int32 timeout) in D:\a\1\s\WebApp\Helpers\OpcSessionHelper.cs:line 937
> 
>    at Microsoft.Azure.IoTSuite.Connectedfactory.WebApp.Helpers.OpcSessionHelper.<GetSessionAsync>d__15.MoveNext() in D:\a\1\s\WebApp\Helpers\OpcSessionHelper.cs:line 198
> 
> --- End of stack trace from previous location where exception was thrown ---
> 
>    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
> 
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
> 
>    at Microsoft.Azure.IoTSuite.Connectedfactory.WebApp.Controllers.BrowserController.<Connect>d__6.MoveNext() in D:\a\1\s\WebApp\Controllers\BrowserController.cs:line 128
>

**Comment by hansgschossmann** [comment]
This is strange. We will take a look.

**Comment by VDanylevskyi** [comment]
Hi
I have the same problem, I can not use publisher. 
I found, that it transforms url of the publisher (replace first part of the url by word publisher):
instead of  `opc.tcp://10.3.100.50:62222/UA/Publisher`
it uses `opc.tcp://publisher.3.100.50:62222/UA/Publisher`

Best regards

**Comment by hansgschossmann** [comment]
@eoursel here a short explanation what happens assuming you are entering "opc.tcp://publisher:62222" in the Cf connect to server text input box and then press the connect button:
- the browser UX code  calls a REST API ("Connect") in the browser controller
- this goes down the OPC UA stack. the stack in Cf is modified to use a transport which is leveraging OPC Proxy to communicate with the OPC server (on our case opc.tcp://publisher:62222)
- the first thing which happens now is that the hostname "publisher" must be resolvable via proxy.
- in your case: have you a proxy running in your CentOS environment as well? If not you should, then the Cf will ask this proxy to resolve the hostname "publisher" and return an IP address and then reports this back to Cf's proxy transport implementation running in the cloud.
- next is that the certificate needs to be validated, but it looks like you are not yet at this step

**Comment by hansgschossmann** [comment]
@VDanylevskyi this is a bug when building the publisher URI (in BrowserController::ConnctToPublisher. I am fixing a few bugs right now and will also address this one. Thanks for pointing it out.

**Comment by eoursel** [comment]
Hi @hansgschossmann 

Thanks for you advice. I succeeded to make it work. I deployed the Azure Iot Connected Factory with the build.ps script from a windows 10 machine and I used a docker compose file to launch the proxy, publisher and test opc ua server on a centos 7 host. 

Nevertheless, I believe we should enhance the configuration of the edge part by providing integration with[ Azure IoT Edge Tool](https://github.com/Azure/iotedgedev), that is to say automate the deployment of the publisher/proxy as IoT Edge modules.
